### PR TITLE
Add new version of icu_pattern crate

### PR DESCRIPTION
**Pull request by sffc**
We need this in so many places that I decided to build it.

I added a new type `NumericPlaceholderPattern` which, as the name suggests, supports only numeric placeholders. It stores all of its data in a string, so it can be stored easily in a data provider and interpolated zero-copy. I took some inspiration from the current `icu_pattern` when designing the API, but my proposed API has fewer generics and instead specialized on this one common use case.

Two things we don't support, and the path to add them:

1. String placeholders. This is easy enough to add as a new type like `StringPlaceholderPattern`.
2. Placeholders with additional context (like datetime patterns). It seems like this is fairly specific to the DateTime use case so I don't know if we want to support them here, but if we do, they can be added as another type like `ContextualPlaceholderPattern`.

The only code I copied from the existing crate was the `Parser` impl, which turns out to be more lines of code than the fresh code that I wrote.

@zbraniecki, if you like this direction, I can just move this code into the existing path in the file tree.

### CODE REVIEW

**Comment by sffc** [comment]
Docs preview: http://icu4x-pr-artifacts.storage.googleapis.com/gha/eb310a34dc12d569ab2cff196256b8088551b9ee/rustdoc/icu_pattern_2/index.html

File: experimental/pattern/src/parser/mod.rs [hunk 1]
```diff
@@ -0,0 +1,665 @@
...
+/// Serveral design decisions have been made that the reader should be aware of when using the API.
+///
+/// ## Zero copy
+///
+/// The parser is intended for runtime use and is optimized for performance and low memory overhad.
+///
+/// Zero copy parsing is a model which allows the parser to produce tokens that are de-facto
+/// slices of the input without ever having to modify the input or copy from it.
+///
+/// In case of ICU patterns that decision brings a trade-off around handling of quoted literals.
```
Comment (robertbastian) [inline experimental/pattern/src/parser/mod.rs:224]: what are "ICU patterns"?

**Comment by sffc** [comment]
@robertbastian The `parser/mod.rs` file is copied verbatim from `utils/pattern`. Does the rest of the PR look good to you?

**Review by Manishearth — APPROVED** [review]
design seems fine overall. didn't look too closely at the parser

**Comment by sffc** [comment]
@zbraniecki Are you OK if I replace `/utils/pattern` with this new code? I will keep your parser but replace the rest of the crate with this code.

**Comment by sffc** [comment]
Discussed with @robertbastian and agreed to make a version that includes a generic for the number of placeholders with potential data model optimizations.

**Comment by sffc** [comment]
Based on feedback from @robertbastian I opened a new PR with a different data model, #4610. I think both data models have pros and cons.

For reference, here is the data model for this PR:

https://icu4x-pr-artifacts.storage.googleapis.com/gha/eb310a34dc12d569ab2cff196256b8088551b9ee/rustdoc/icu_pattern_2/struct.NumericPlaceholderPattern.html#pattern-string-encoding-details

I have not run performance benchmarks and can do so if that is the deciding factor. As far as data size, I expect the model implemented in this PR to be slightly smaller than the one implemented in #4610.

**Comment by sffc** [comment]
I'm quite happy now with #4610 so I will close this PR.

**Pull request closed without merging**